File: admin/credential_status.go
package admin

// Credential display state is intentionally derived from the orthogonal
// persistence axes. Runtime routing keeps using the original columns; this
// value is the single operator-facing state used by admin APIs and UI.
type CredentialDisplayState string

const (
	CredentialStateActive         CredentialDisplayState = "active"
	CredentialStateCooling        CredentialDisplayState = "cooling"
	CredentialStateDegraded       CredentialDisplayState = "degraded"
	CredentialStateRateLimited    CredentialDisplayState = "rate_limited"
	CredentialStateUnreachable    CredentialDisplayState = "unreachable"
	CredentialStateAuthFailed     CredentialDisplayState = "auth_failed"
	CredentialStateSuspended      CredentialDisplayState = "suspended"
	CredentialStateQuotaExhausted CredentialDisplayState = "quota_exhausted"
	CredentialStateDisabled       CredentialDisplayState = "disabled"
	CredentialStateDeleted        CredentialDisplayState = "deleted"
	CredentialStateUnknown        CredentialDisplayState = "unknown"
)

var credentialStatuses = map[string]struct{}{
	"active": {}, "cooling": {}, "degraded": {}, "quarantine": {},
	"quota_expired": {}, "disabled": {}, "deleted": {},
}

func isCredentialStatus(value string) bool {
	_, ok := credentialStatuses[value]
	return ok
}

type credentialStateInput struct {
	Status          string
	LifecycleStatus string
	Availability    string
	QuotaState      string
	HealthStatus    string
	ManualDisabled  bool
}

type credentialDisplayState struct {
	State  CredentialDisplayState
	Reason string
}

// deriveCredentialDisplayState applies the documented precedence from most
// terminal/operator-controlled to least severe transient condition.
func deriveCredentialDisplayState(in credentialStateInput) credentialDisplayState {
	if in.Status == "deleted" || in.LifecycleStatus == "retired" {
		return credentialDisplayState{CredentialStateDeleted, "deleted"}
	}
	if in.ManualDisabled || in.Status == "disabled" || in.LifecycleStatus == "disabled" {
		return credentialDisplayState{CredentialStateDisabled, "manual_disabled"}
	}
	if in.Availability == "auth_failed" || in.HealthStatus == "unreachable" && in.Availability == "auth_failed" {
		return credentialDisplayState{CredentialStateAuthFailed, "auth_failed"}
	}
	if in.QuotaState == "balance_exhausted" || in.QuotaState == "permanently_exhausted" || in.QuotaState == "periodic_exhausted" || in.Status == "quota_expired" {
		return credentialDisplayState{CredentialStateQuotaExhausted, in.QuotaState}
	}
	if in.Availability == "suspended" {
		return credentialDisplayState{CredentialStateSuspended, "suspended"}
	}
	if in.Availability == "rate_limited" {
		return credentialDisplayState{CredentialStateRateLimited, "rate_limited"}
	}
	if in.Availability == "unreachable" || in.HealthStatus == "unreachable" {
		return credentialDisplayState{CredentialStateUnreachable, "unreachable"}
	}
	if in.Availability == "cooling" || in.Status == "cooling" {
		return credentialDisplayState{CredentialStateCooling, "cooling"}
	}
	if in.Availability == "degraded" || in.Status == "degraded" || in.HealthStatus == "warning" {
		return credentialDisplayState{CredentialStateDegraded, "degraded"}
	}
	if in.Status == "active" || in.LifecycleStatus == "active" || in.Availability == "ready" || in.HealthStatus == "healthy" {
		return credentialDisplayState{CredentialStateActive, "active"}
	}
	return credentialDisplayState{CredentialStateUnknown, "unknown"}
}

File: web/src/utils/credentialStatus.ts
export const CREDENTIAL_DISPLAY_STATES = [
  'active', 'cooling', 'degraded', 'rate_limited', 'unreachable',
  'auth_failed', 'suspended', 'quota_exhausted', 'disabled', 'deleted', 'unknown',
] as const

export type CredentialDisplayState = typeof CREDENTIAL_DISPLAY_STATES[number]

export interface CredentialStatusLike {
  effective_state?: string | null
  effective_reason?: string | null
  status?: string | null
  lifecycle_status?: string | null
  availability_state?: string | null
  health_status?: string | null
  quota_state?: string | null
  manual_disabled?: boolean | null
}

export function credentialDisplayState(value: CredentialStatusLike): CredentialDisplayState {
  const explicit = value.effective_state as CredentialDisplayState | undefined
  if (explicit && (CREDENTIAL_DISPLAY_STATES as readonly string[]).includes(explicit)) return explicit
  if (value.status === 'deleted' || value.lifecycle_status === 'retired') return 'deleted'
  if (value.manual_disabled || value.status === 'disabled' || value.lifecycle_status === 'disabled') return 'disabled'
  if (value.availability_state === 'auth_failed') return 'auth_failed'
  if (['balance_exhausted', 'permanently_exhausted', 'periodic_exhausted'].includes(value.quota_state ?? '')) return 'quota_exhausted'
  if (value.availability_state === 'suspended') return 'suspended'
  if (value.availability_state === 'rate_limited') return 'rate_limited'
  if (value.availability_state === 'unreachable' || value.health_status === 'unreachable') return 'unreachable'
  if (value.availability_state === 'cooling' || value.status === 'cooling') return 'cooling'
  if (value.availability_state === 'degraded' || value.status === 'degraded' || value.health_status === 'warning') return 'degraded'
  if (value.status === 'active' || value.lifecycle_status === 'active' || value.availability_state === 'ready' || value.health_status === 'healthy') return 'active'
  return 'unknown'
}

export const credentialDisplayStateLabel: Record<CredentialDisplayState, string> = {
  active: 'Active', cooling: 'Cooling', degraded: 'Degraded', rate_limited: 'Rate limited',
  unreachable: 'Unreachable', auth_failed: 'Auth failed', suspended: 'Suspended',
  quota_exhausted: 'Quota exhausted', disabled: 'Disabled', deleted: 'Deleted', unknown: 'Unknown',
}
